### ISSUE TITLE
revert command causing issue: Thread 'main' panicked at 'Best block header must always exist', src/libcore/option.rs:1008

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#1347]
I did `substrate revert --chain=mydev` and got 
```
2019-01-04 14:29:03 Reverted 256 blocks. Best: #90996 (0x3a45…e4e6)
```

then

```
$ ./target/debug/substrate --chain dev.json
2019-01-04 14:30:12 Substrate Node
2019-01-04 14:30:12   version 0.9.1-f5863909-x86_64-macos
2019-01-04 14:30:12   by Parity Technologies, 2017, 2018
2019-01-04 14:30:12 Chain specification: My DEV
2019-01-04 14:30:12 Node name: sedate-existence-5917
2019-01-04 14:30:12 Roles: FULL

====================

stack backtrace:
   0:        0x1111dfb44 - backtrace::backtrace::trace::h12f7306b347914bd
                        at /Users/bryanchen/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/backtrace/mod.rs:42
   1:        0x1111da9dc - backtrace::capture::Backtrace::new_unresolved::h2d0fecf94b11e6b1
                        at /Users/bryanchen/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:88
   2:        0x1111da93e - backtrace::capture::Backtrace::new::h517f18887a64b9f0
                        at /Users/bryanchen/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:63
   3:        0x10f25af27 - substrate_cli::panic_hook::panic_hook::h9ac81556be38b044
                        at /Users/bryanchen/projects/substrate/core/cli/src/panic_hook.rs:51
   4:        0x10f2474e8 - core::ops::function::Fn::call::h454cfe96ff54e1a1
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/ops/function.rs:78
   5:        0x111537ff1 - std::panicking::rust_panic_with_hook::h8a9190241e1c3560
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:495
   6:        0x111537a3c - std::panicking::continue_panic_fmt::h9deb5a55af7ba8bb
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:398
   7:        0x111537928 - _rust_begin_unwind
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:325
   8:        0x11154e321 - core::panicking::panic_fmt::h63f09b1c8394e778
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/panicking.rs:95
   9:        0x11154e398 - core::option::expect_failed::hde5db0fb33217d93
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/option.rs:1008
  10:        0x10eaa6621 - <core::option::Option<T>>::expect::h9067027a41a4b63f
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/option.rs:322
  11:        0x10ee7d5c8 - <substrate_client::client::Client<B, E, Block, RA>>::best_block_header::h1626c26f25d6ee0b
                        at /Users/bryanchen/projects/substrate/core/client/src/client.rs:821
  12:        0x10edffb29 - <substrate_service::Service<Components>>::new::hf29c16c8c1c0a6bd
                        at /Users/bryanchen/projects/substrate/core/service/src/lib.rs:157
  13:        0x10e8795fc - <substrate_service::components::FullComponents<Factory>>::new::h6d2ba77f5f023aae
                        at /Users/bryanchen/projects/substrate/core/service/src/components.rs:398
  14:        0x10ee5bad4 - <node_cli::service::Factory as substrate_service::components::ServiceFactory>::new_full::{{closure}}::hd31e13e6b0986ee4
                        at /Users/bryanchen/projects/substrate/node/cli/src/service.rs:72
  15:        0x10e99d541 - <node_cli::service::Factory as substrate_service::components::ServiceFactory>::new_full::h9a90f0d0df7a2fd1
                        at /Users/bryanchen/projects/substrate/<::substrate_service::construct_service_factory macros>:50
  16:        0x10e6a9c94 - node_cli::run::h6215dbefc70f0e6a
                        at /Users/bryanchen/projects/substrate/node/cli/src/lib.rs:141
  17:        0x10e63bd6d - substrate::run::h9daa42d50a2f2601
                        at /Users/bryanchen/projects/substrate/node/src/main.rs:63
  18:        0x10e63bd96 - substrate::main::h0e516cbb633fad37
                        at /Users/bryanchen/projects/substrate/<::error_chain::quick_main::quick_main macros>:4
  19:        0x10e67ebd1 - std::rt::lang_start::{{closure}}::h8f3e7c02a68d3743
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/rt.rs:74
  20:        0x1115378a7 - std::panicking::try::do_call::h59d1e5d7439c66ee
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:310
  21:        0x1115479ce - ___rust_maybe_catch_panic
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libpanic_unwind/lib.rs:102
  22:        0x1115382fa - std::rt::lang_start_internal::h3e2807095f24f52e
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/rt.rs:58
  23:        0x10e67ebb1 - std::rt::lang_start::h5f4f918f1f8bf070
                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/rt.rs:74
  24:        0x10e63bf81 - _main

Thread 'main' panicked at 'Best block header must always exist', src/libcore/option.rs:1008

This is a bug. Please report it at:

    https://github.com/paritytech/polkadot/issues/new
```

**Comment by JKingdom** [comment]
@xlc 
I seems that I also meet this, so I provide this issue #1342 and pr #1341 trying to fix this problem.
the parity teem in holiday now,
maybe we would wait a long time to fix this, hhh

**Comment by xlc** [comment]
I can confirm #1341 is able to revert correctly.
It is unable to fix the broken db caused by the broken revert command so I need to re-sync the chain data...

@JKingdom 加个微信？微信号 xlc0212

**Comment by xlc** [comment]
Fixed by #1341
